xen/arch/x86/tpm.c

```diff
@@ -978,11 +971,150 @@ void tpm_hash_extend(unsigned loc, unsigned pcr, uint8_t *buf, unsigned size,
+        TPM_PANIC("SLRT is missing Intel-specific information!\n");
+
+    policy = (struct slr_entry_policy *)
+        slr_next_entry_by_tag(slrt, NULL, SLR_ENTRY_DRTM_POLICY);
+    if (policy == NULL)
+        TPM_PANIC("SLRT is missing DRTM policy!\n");
+
+    /* SLRT policy entry must be the first one (so measuring order matches
+     * policy order) and point to SLRT. */
+    /* XXX: are newer revisions allowed? */
+    if ( policy->revision != SLR_POLICY_REVISION )
+        TPM_PANIC("DRTM policy in SLRT is of unsupported revision: %#04x!\n",
+                  slrt->revision);
+    if ( policy->nr_entries == 0 )
+        TPM_PANIC("DRTM policy in SLRT is empty!\n");
+
+    policy_entry = (struct slr_policy_entry *)
+        ((uint8_t *)policy + sizeof(*policy));
+    if ( policy_entry->entity_type != SLR_ET_SLRT )
+        TPM_PANIC("First entry of DRTM policy in SLRT is empty!\n");
+    if ( policy_entry->entity != os_mle->slrt )
+        TPM_PANIC("SLRT entry of DRTM policy points to wrong location!\n");
+
+    return policy;
+
+#undef TPM_PANIC
+}
+
 #ifdef __EARLY_TPM__
-void __stdcall tpm_extend_mbi(uint32_t *mbi)
+static void tpm_extend_mbi(uint32_t *mbi)
 {
     /* MBI starts with uint32_t total_size. */
     tpm_hash_extend(DRTM_LOC, DRTM_DATA_PCR, (uint8_t *)mbi, *mbi,
                     TXT_EVTYPE_MBI, NULL, 0);
 }
+
+static void tpm_extend_slrt(void)
+{
+    struct txt_os_mle_data *os_mle;
+    struct slr_table *slrt;
+    struct slr_entry_policy *policy;
+    struct slr_policy_entry *policy_entry;
+
+    os_mle = txt_os_mle_data_start(__va(read_txt_reg(TXTCR_HEAP_BASE)));
+    slrt = __va(os_mle->slrt);
+
+    /* All errors are handled by doing nothing here, the checks will be repeated
+     * and handled for real once more things are initialized. */
+    policy = tpm_get_policy();
+    if ( policy == NULL )
+        return;
+
+    policy_entry = (struct slr_policy_entry *)
+        ((uint8_t *)policy + sizeof(*policy));
+    tpm_hash_extend(DRTM_LOC, policy_entry->pcr, (uint8_t *)slrt,
+                    policy_entry->size, TXT_EVTYPE_SLRT, NULL, 0);
+}
+
+void __stdcall tpm_early_extend(uint32_t *mbi)
+{
+    tpm_extend_mbi(mbi);
+    tpm_extend_slrt();
+}
+#else
+void tpm_take_measurements(void)
+{
+    struct txt_os_mle_data *os_mle;
+    struct slr_table *slrt;
+    struct slr_entry_policy *policy;
+    struct slr_policy_entry *policy_entry;
+    uint16_t i;
+
+    os_mle = txt_os_mle_data_start(__va(read_txt_reg(TXTCR_HEAP_BASE)));
+    slrt = __va(os_mle->slrt);
+
+    policy = tpm_get_policy();
+    policy_entry = (struct slr_policy_entry *)
+        ((uint8_t *)policy + sizeof(*policy));
+
+    for (i = 1; i < policy->nr_entries; i++) {
+        uint64_t start = policy_entry[i].entity;
+        uint64_t size = policy_entry[i].size;
+        uint32_t event_type = TXT_EVTYPE_UNKNOWN;
+
+        switch (policy_entry[i].entity_type) {
+        case SLR_ET_SLRT:
```

Comment:
This needs special treatment, described in Appendix A of Secure Launch Specification. SLRT contains (among other things) addresses that may change between boots, these are not measured.

Reply:
That appendix is actually about DRTM, which is one of multiple SLRT entries that contains pointers (there are DCE, DL handler, bootloader context and probably other fields).  Hashing of SLRT then also needs to be done by extending hash with each entry in turn taking pointer values into account and "recursively" processing DRTM.

As mentioned in https://github.com/TrenchBoot/grub/pull/14#discussion_r1388220696, probably need some consistent policy on such things.

Reply:
Fair point. Perhaps we should organize a (public? not sure how many people would be interested in attending) meeting for discussing everything that specification lacks, but there are still many issues being discovered so maybe now would be too soon.

Reply:
New way of handling SLRT as per https://github.com/TrenchBoot/xen/pull/7#issuecomment-1821023039: https://github.com/TrenchBoot/xen/compare/d820e3dac46047e7e81db6fdc30d72041e9d208e..6ec1f7b7ae9af57ca6abbb7729255aaf14fbcfa7#diff-5861d423276bd98c31fb2a501569110297cbe6488842e0359435a3af795c24aeR1113-R1127